// poker-odds-calculator/dist/CardGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Card_1 = require("./Card");
const _ = require("lodash");
class CardGroup extends Array {
    constructor() {
        super();
    }
    static fromString(s) {
        const tmp = s.replace(/[^a-z0-9]/gi, '');
        if (tmp.length % 2 !== 0) {
            throw new Error(`Invalid cards: ${s}`);
        }
        const cardgroup = new CardGroup();
        for (let i = 0; i < tmp.length; i = i + 2) {
            cardgroup.push(Card_1.Card.fromString(tmp.substring(i, i + 2)));
        }
        return cardgroup;
    }
    static fromCards(cards) {
        const cardgroup = new CardGroup();
        for (const card of cards) {
            cardgroup.push(card);
        }
        return cardgroup;
    }
    contains(c) {
        for (const card of this) {
            if (card.equals(c)) {
                return true;
            }
        }
        return false;
    }
    toString() {
        return '' + this.join(' ');
    }
    sortCards(type) {
        const sorted = _.orderBy(this, ['rank', 'suit'], [type, type]);
        this.splice.apply(this, [0, this.length].concat(sorted));
    }
    concat(cardgroup) {
        const ret = new CardGroup();
        for (const card of this) {
            ret.push(card);
        }
        for (const card of cardgroup) {
            ret.push(card);
        }
        return ret;
    }
    countBy(type) {
        return _.countBy(this, type);
    }
}
exports.CardGroup = CardGroup;
